### PR TITLE
Feature/mongo datastore

### PR DESCRIPTION
**Pull request by svevang**
This PR replaces redis with mongo as the datastore. It will allow the game to scale beyond the in-memory limits of redis.

mocha and jasmie tests are passing. The zombie acceptance tests are still busted, and will need to be fixed up in a later PR.

### CODE REVIEW

File: lib/persistence/mongo/Base.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    idAttribute: "_id",
+
+    create: function(attributes, options){
+
+      if(!this.collection_name){
+        throw new Error("need the collection_name attribute on the model!")
+      }
+
+      // 24 character id string, as opposed to using mongo's builtin ObjectID
+      this.set('_id', crypto.randomBytes(12).toString('hex'))
```
Comment (michaelkirk) [inline lib/persistence/mongo/Base.js]: Why are we using crypto here instead of the built in ID?

File: lib/persistence/mongo/Board.js [hunk 2]
```diff
@@ -32,16 +33,16 @@ module.exports = function(weiqi){
...
       });
     },
-    find_white_player_id: function() {
-      return weiqi.Player.id_for_board_id_and_color(this.id, "white");
+    find_white_player: function() {
+      return weiqi.Player.find_by_board_id_and_color(this.id, "white");
     },
-    find_black_player_id: function() {
-      return weiqi.Player.id_for_board_id_and_color(this.id, "black");
+    find_black_player: function() {
```
Comment (michaelkirk) [inline lib/persistence/mongo/Board.js]: This is nicer.

File: lib/persistence/mongo/Base.js (hunk 1 above)
Comment (svevang) [inline lib/persistence/mongo/Base.js]: I think the main reason is that the mongo [ObjectId](http://docs.mongodb.org/manual/reference/object-id/) is not a simple string in the Node Mongo library. You have to explicitly serialize and deserialize the ObjectId instance to a string for Backbone to consume.  It was a little bit cumbersome doing this for the crud operations

The other aspect of this, is that for some operations like `create` and `read` we are returning the javascript object result of the lookup, not the backbone model instance. So those cases we would have to intercept the ObjectID's anyways and turn them into strings. 

It seems easier to consider the id to be a unique string and that's it.

Comment (svevang) [inline lib/persistence/mongo/Base.js]: ahh, and also you could easily predict the next id in the sequence. The boards were being created so quickly that the timestamps overlapped:

      weiqi> console.log(ObjectID(), ObjectID())
    54d26136aaebe2dc0b0420da 54d26136aaebe2dc0b0420db

Comment (michaelkirk) [inline lib/persistence/mongo/Base.js]: I'm not really familiar with mongo, but it seems like we're breaking a
convention here, and I'm just trying to decide if it's the kind of
convention that we should be sticking to.

File: lib/persistence/mongo/Base.js [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    pk: function(){
+      return this.get('_id');
+    },
+
+    idAttribute: "_id",
+
+    create: function(attributes, options){
+
+      if(!this.collection_name){
```
Comment (michaelkirk) [inline lib/persistence/mongo/Base.js]: The purpose of `collection_name` in mongo seems to be the same as `name` in redis. I don't really care which one we use, but I think you should get rid of one or the other.

File: lib/persistence/mongo/Base.js (hunk 1 above)
Comment (svevang) [inline lib/persistence/mongo/Base.js]: My sense (and I could be wrong) is that ObjectID's are only partly conventional. Where you would abandon them is when you have slightly custom requirements for ids (e.g. they are not just a key to lookup and join documents).

See e.g. meteor's coice to use strings:
https://groups.google.com/d/msg/meteor-talk/f-ljBdZOwPk/oQYZQxCAKN8J
https://groups.google.com/d/msg/meteor-talk/f-ljBdZOwPk/oQYZQxCAKN8J

File: lib/persistence/mongo/Base.js (hunk 3 above)
Comment (svevang) [inline lib/persistence/mongo/Base.js]: nice catch!, added d7facb4

**Comment by michaelkirk** [comment]
:boom:

**Comment by michaelkirk** [comment]
Well the exciting new is that this deployed. Less exciting is that it's erroring. Seems to be still connecting to local mongo port.

I've fixed the defaults and configured heroku but still nothing. Gotta run to lunch.

**Comment by michaelkirk** [comment]
Woohoo! I think I  fixed it with c6a5f16ac647c6429623968e9640fa9e98bb6e2c.

check out http://go.jackpine.me

**Comment by svevang** [comment]
yes!, There is a slight bug in the formatting of the invite, possible fix here:
https://github.com/michaelkirk/weiqi/commit/4da3da34542e4b52c8a5fd9c3712f6f2d1d332d4

**Pull request merged**